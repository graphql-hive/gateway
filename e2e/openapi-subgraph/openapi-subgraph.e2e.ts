import { createTenv, Service } from '@internal/e2e';
import { beforeAll, describe, expect, it } from 'vitest';

describe('openapi-subgraph', () => {
  let OASService: Service;
  let GQLService: Service;
  const { service, composeWithMesh, gateway } = createTenv(__dirname);
  beforeAll(async () => {
    OASService = await service('OAS');
    GQLService = await service('GQL');
  });
  function replaceDockerHostNamesBack(sdl?: string) {
    return sdl?.replaceAll('172.17.0.1', 'localhost');
  }
  it('exposes the SDL correctly', async () => {
    const { result, output } = await composeWithMesh({
      services: [OASService, GQLService],
      maskServicePorts: true,
      args: ['--subgraph', 'OAS'],
      output: 'graphql',
    });
    const { execute } = await gateway({
      subgraph: output,
    });
    const sdlQuery = /* GraphQL */ `
      query {
        _service {
          sdl
        }
      }
    `;
    const queryResult = await execute({
      query: sdlQuery,
    });
    expect(queryResult?.errors).toBeFalsy();
    expect(replaceDockerHostNamesBack(queryResult?.data?._service?.sdl)).toBe(
      replaceDockerHostNamesBack(result),
    );
  });
  it('resolves entitites correctly', async () => {
    const { execute } = await gateway({
      subgraph: {
        with: 'mesh',
        services: [OASService, GQLService],
        subgraphName: 'OAS',
      },
    });
    const entitiesQuery = /* GraphQL */ `
      query {
        _entities(representations: [{ __typename: "User", id: 1 }]) {
          __typename
          ... on User {
            id
            name
          }
        }
      }
    `;
    const queryResult = await execute({
      query: entitiesQuery,
    });
    expect(queryResult?.errors).toBeFalsy();
    expect(queryResult?.data).toEqual({
      _entities: [
        {
          __typename: 'User',
          id: '1',
          name: 'Alice',
        },
      ],
    });
  });
  it('encapsulates the queries correctly', async () => {
    const { execute } = await gateway({
      subgraph: {
        with: 'mesh',
        services: [OASService, GQLService],
        subgraphName: 'GQL',
      },
    });
    const encapsulatedQuery = /* GraphQL */ `
      {
        gql {
          books {
            id
            title
          }
        }
      }
    `;
    const queryResult = await execute({
      query: encapsulatedQuery,
    });
    expect(queryResult?.errors).toBeFalsy();
    expect(queryResult?.data).toEqual({
      gql: {
        books: [
          {
            id: '1',
            title: 'Book 1',
          },
          {
            id: '2',
            title: 'Book 2',
          },
          {
            id: '3',
            title: 'Book 3',
          },
        ],
      },
    });
  });
});
