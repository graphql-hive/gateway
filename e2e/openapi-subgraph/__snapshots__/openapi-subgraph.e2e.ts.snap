// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi-subgraph generates a subgraph SDL from an OpenAPI spec 1`] = `
"schema @link(url: "https://specs.apollo.dev/link/v1.0") @link(url: "https://specs.apollo.dev/federation/v2.3", import: ["@key", "@composeDirective"]) @link(url: "https://the-guild.dev/graphql/mesh/spec/v1.0", import: ["@merge", "@statusCodeTypeName", "@httpOperation", "@transport", "@source", "@extraSchemaDefinitionDirective"]) @composeDirective(name: "@merge") @composeDirective(name: "@statusCodeTypeName") @composeDirective(name: "@httpOperation") @composeDirective(name: "@transport") @composeDirective(name: "@source") @composeDirective(name: "@extraSchemaDefinitionDirective") {
  query: Query
}

directive @statusCodeTypeName(subgraph: String, typeName: String, statusCode: String) repeatable on UNION

directive @httpOperation(subgraph: String, path: String, operationSpecificHeaders: [[String]], httpMethod: HTTPMethod, isBinary: Boolean, requestBaseBody: ObjMap, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap, jsonApiFields: Boolean, queryStringOptions: ObjMap) repeatable on FIELD_DEFINITION

directive @transport(subgraph: String, kind: String, location: String, headers: [[String]], queryStringOptions: ObjMap, queryParams: [[String]]) repeatable on SCHEMA

directive @merge(subgraph: String!, argsExpr: String, keyArg: String, keyField: String, key: [String], additionalArgs: String) repeatable on FIELD_DEFINITION

type Query @extraSchemaDefinitionDirective(directives: {transport: [{subgraph: "Test", kind: "rest", location: "http://localhost:<Test_port>"}]}) {
  users: [User] @httpOperation(subgraph: "Test", path: "/users", operationSpecificHeaders: [["accept", "application/json"]], httpMethod: GET)
  user(id: String!): user_response @httpOperation(subgraph: "Test", path: "/users/{args.id}", operationSpecificHeaders: [["accept", "application/json"]], httpMethod: GET) @merge(subgraph: "Test", keyField: "id", keyArg: "id")
}

type User @key(fields: "id") {
  id: String!
  name: String!
}

union user_response @statusCodeTypeName(subgraph: "Test", statusCode: "200", typeName: "User") @statusCodeTypeName(subgraph: "Test", statusCode: "404", typeName: "user_404_response") = User | user_404_response

type user_404_response {
  message: String!
}

enum HTTPMethod {
  GET
  HEAD
  POST
  PUT
  DELETE
  CONNECT
  OPTIONS
  TRACE
  PATCH
}

scalar ObjMap

directive @source(name: String!, type: String, subgraph: String!) repeatable on SCALAR | OBJECT | FIELD_DEFINITION | ARGUMENT_DEFINITION | INTERFACE | UNION | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

scalar _DirectiveExtensions

directive @extraSchemaDefinitionDirective(directives: _DirectiveExtensions) repeatable on OBJECT
"
`;
